File: atcoder/abc329_f.py
# https://atcoder.jp/contests/abc329/tasks/abc329_f

import sys
# input = sys.stdin.buffer.readline
# def input(): return sys.stdin.readline().rstrip()
# sys.setrecursionlimit(10 ** 7)

N, Q = map(int, (input().split()))
C = list(map(int, (input().split())))

box = [set([C[i]]) for i in range(N)]
for qi in range(Q):
    a, b = map(int, input().split())
    a -= 1; b -= 1
    if len(box[b]) > len(box[a]):
        box[b] |= box[a]
    else:
        box[a] |= box[b]
        box[a], box[b] = box[b], box[a]
    box[a] = set()
    print(len(box[b]))





# N, Q = map(int, (input().split()))
# C = list(map(int, (input().split())))

# box = [set() for _ in range(N)]
# for i in range(N):
#     box[i].add(C[i])

# idx = list(range(N))

# for qi in range(Q):
#     a, b = map(int, input().split())
#     a -= 1; b -= 1
#     if len(box[idx[b]]) >= len(box[idx[a]]):
#         box[idx[b]] |= box[idx[a]]
#         box[idx[a]] = set()
#     else:
#         idx[a], idx[b] = idx[b], idx[a]
#         box[idx[b]] |= box[idx[a]]
#         box[idx[a]] = set()
#     print(len(box[idx[b]]))
#     # print(*idx, file=sys.stderr)

File: atcoder/abc329_f.cpp
// https://atcoder.jp/contests/abc329/tasks/abc329_f
#include<iostream>
// #include<algorithm>
// #include<cmath>
// #include <ctime>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
// #include <atcoder/all>
// using namespace atcoder;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
#define maxs(x,y) (x = max(x,y))
#define mins(x,y) (x = min(x,y))
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
	int n, q;
	cin >> n >> q;
    // cout << n << ' ' << q << endl;
    vector<set<int>> box(n, set<int>());
    rep(i,n) {
        int c;
        cin >> c;
        box[i].insert(c);
    }

    rep(i,q) {
        int a, b;
        cin >> a >> b;
        --a, --b;
        // cout << a << ' ' << b << endl;
        if (box[a].size() < box[b].size()) {
            for (int x: box[a]) box[b].insert(x);
        } else {
            for (int x: box[b]) box[a].insert(x);
            swap(box[a],box[b]);
        }
        box[a].clear();
        cout << box[b].size() << endl;
    }

	return 0;
}
